mountpoint-s3/tests/reftests/harness.rs

```diff
@@ -143,189 +287,248 @@ impl Harness {
-            Name("a".to_string()),
-            TreeNode::Directory(BTreeMap::from([
-                (Name("a/".to_string()), TreeNode::File(Content(0, FileSize::Small(0)))),
-                (Name("a".to_string()), TreeNode::File(Content(0, FileSize::Small(0)))),
-            ])),
-        )])),
-        CheckType::FullTree,
-        0,
-    )
-}
+/// Mutation tests that run a sequence of mutations against a file system and check equivalence to
+/// the reference model.
+mod mutations {
+    use super::*;
+    use proptest::collection::vec;
+
+    fn run_test(initial_tree: TreeNode, ops: Vec<Op>, readdir_limit: usize) {
+        let test_prefix = Prefix::new("test_prefix/").expect("valid prefix");
+        let config = S3FilesystemConfig {
+            readdir_size: 5,
+            ..Default::default()
+        };
+        let (client, fs) = make_test_filesystem("harness", &test_prefix, config);
+
+        let namespace = flatten_tree(initial_tree);
+        for (key, object) in namespace.iter() {
+            client.add_object(&format!("{test_prefix}{key}"), object.to_mock_object());
+        }
 
-#[test]
-fn random_tree_regression_directory_shadow_lookup() {
-    run_test(
-        TreeNode::Directory(BTreeMap::from([(
-            Name("a".to_string()),
-            TreeNode::Directory(BTreeMap::from([
-                (Name("a/".to_string()), TreeNode::File(Content(0, FileSize::Small(0)))),
-                (Name("a".to_string()), TreeNode::File(Content(0, FileSize::Small(0)))),
-            ])),
-        )])),
-        CheckType::SinglePath { path_index: 1 },
-        0,
-    )
+        let reference = build_reference(namespace);
+
+        let mut harness = Harness::new(fs, reference, readdir_limit);
+
+        futures::executor::block_on(harness.run(ops));
+    }
+
+    proptest! {
+        #![proptest_config(ProptestConfig {
+            failure_persistence: None,
+            .. ProptestConfig::default()
+        })]
+
+        #[test]
+        fn reftest_random_tree(tree in gen_tree(5, 100, 5, 20), readdir_limit in 0..10usize, ops in vec(any::<Op>(), 1..10)) {
+            run_test(tree, ops, readdir_limit);
+        }
+    }
+
+    #[test]
+    fn regression_basic() {
+        run_test(
+            TreeNode::Directory(BTreeMap::from([(
+                Name("-".to_string()),
+                TreeNode::Directory(BTreeMap::from([(
+                    Name("-".to_string()),
+                    TreeNode::File(FileContent(0, FileSize::Small(0))),
+                )])),
+            )])),
+            vec![
+                Op::WriteFile(
+                    "a".to_string(),
+                    DirectoryIndex(0),
+                    FileContent(0x0a, FileSize::Small(50)),
+                ),
+                Op::WriteFile(
+                    "b".to_string(),
+                    DirectoryIndex(1),
+                    FileContent(0x0b, FileSize::Small(10)),
+                ),
+            ],
+            0,
+        );
+    }
```

Comment:
Just to confirm understanding on this one (and I see we already had something similar for RO), we're just checking that the following filesystem behaves as expected so that we can be sure we didn't break some basic case (without needing to worry about the trees that proptest will generate)?

```
/ (dir)
/-/ (dir)
/-/- (file)
```

becomes...


```
/ (dir)
/-/ (dir)
/a (file)
/-/- (file)
/-/b (file)
```

Reply:
Yeah exactly, it's just a useful sanity check. It's most helpful when working on the tests themselves, because it's a lot easier to debug a small test like this than to wait for proptest to do its thing.